tests/test_steps.py

```diff
@@ -0,0 +1,376 @@
+            ]
+        ).run()
+        controller.emit([1, 2, 3])
+        controller.terminate()
+        controller.await_termination()
+    except FlowError:
+        fail("Assert failed unexpectedly", False)
+
+
+def test_assert_none_of():
+    try:
+        controller = build_flow(
+            [
+                Source(),
+                Assert().match_none_of([[1, 2, 3], [4, 5, 6], [7, 8, 9]])
+            ]
+        ).run()
+        controller.emit([10, 11, 12])
+        controller.terminate()
+        controller.await_termination()
+    except FlowError:
+        fail("Assert failed unexpectedly", False)
+
+    try:
+        controller = build_flow(
+            [
+                Source(),
+                Assert().match_none_of([[1, 2, 3], [4, 5, 6], [7, 8, 9]])
+            ]
+        ).run()
+        controller.emit([1, 2, 3])
+        controller.terminate()
+        controller.await_termination()
+        fail("Assert not failing", False)
+    except FlowError:
+        pass
+
+
+def test_sample_by_count():
+    controller = build_flow(
+        [
+            Source(),
+            Assert().exactly(5),
+            Sample(5),
+            Assert().exactly(1).match_exactly([1]),
+        ]
+    ).run()
+
+    controller.emit(1)
+    controller.emit(2)
+    controller.emit(3)
+    controller.emit(4)
+    controller.emit(5)
+    controller.terminate()
+    controller.await_termination()
+
+
+def test_flatten():
+    controller = build_flow(
+        [
+            Source(),
+            Flatten(),
+            Assert().match_exactly([1, 2, 3, 4, 5, 6])
+        ]
+    ).run()
+
+    controller.emit([[1, 2], [3, 4], [5, 6]])
+    controller.terminate()
+    controller.await_termination()
+
+
+def test_foreach():
+    class EventRegistry:
+        def __init__(self):
+            self.events = set()
+
+        def __call__(self, event: Event):
+            self.events.add(event.id)
+            return event
+
+    er = EventRegistry()
+    controller = build_flow(
+        [
+            Source(),
+            Map(er, full_event=True),
+            ForEach(lambda x: None),
+            Assert(full_event=True).each_event(lambda event: event.id in er.events),
+        ]
+    ).run()
+
+    controller.emit(1)
+    controller.emit(1)
+    controller.emit(1)
+    controller.emit(1)
+    controller.emit(1)
+    controller.emit(1)
+
+    controller.terminate()
+    controller.await_termination()
+
+
+def test_partition():
+    dividable_by_two = {2, 4, 6}
+    not_dividable_by_two = {1, 3, 5}
+
+    def check_partition(event: Event):
+        first = event.body.left
+        second = event.body.right
+
+        if first is not None:
+            return first in dividable_by_two and first not in not_dividable_by_two and second is None
+        else:
+            return second in not_dividable_by_two and second not in dividable_by_two
+
+    controller = build_flow(
+        [
+            Source(),
+            Assert().exactly(6),
```

Comment:
Is this testing `Source`?

Reply:
Not really, it's more like "I'm assuring the input of this step"